pkg/backporter/httpcache.go

```diff
@@ -0,0 +1,145 @@
+package backporter
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	xFromCache       = "X-From-Cache"
+	cacheRefreshFreq = 10 * time.Minute
+)
+
+type bugzillaCache struct {
+	lock  sync.Mutex
+	cache map[string][]byte
+}
+
+func (bc *bugzillaCache) get(key string) ([]byte, bool) {
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	cachedVal, ok := bc.cache[key]
+	return cachedVal, ok
+}
+
+func (bc *bugzillaCache) set(key string, respBytes []byte) {
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	bc.cache[key] = respBytes
+}
+
+func newBugzillaCache() *bugzillaCache {
+	return &bugzillaCache{cache: map[string][]byte{}}
+}
+
+// cachingTransport is an implementation http.RoundTripper
+// which first checks for cached values
+type cachingTransport struct {
+	cache     *bugzillaCache
+	transport http.RoundTripper
+}
+
+// RoundTrip will first check if there are any cached responses and return that
+// if not it will make an HTTP call using the DefaultTransport
+func (t *cachingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.Method != "GET" {
+		return t.transport.RoundTrip(req)
+	}
+	var err error
+	var resp *http.Response
+	g := new(errgroup.Group)
+	g.Go(func() error {
+		resp, err = t.transport.RoundTrip(req)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNotFound {
+			body, err := httputil.DumpResponse(resp, true)
+			if err != nil {
+				return fmt.Errorf("err while serializing response to cache: %w", err)
+			}
+			t.cache.set(req.URL.String(), body)
+		}
+
+		return nil
+	})
+
+	if cachedVal, isCached := t.cache.get(req.URL.String()); isCached {
+		b := bytes.NewBuffer(cachedVal)
+		cachedResp, err := http.ReadResponse(bufio.NewReader(b), req)
+		if err != nil {
+			return nil, err
+		}
+		cachedResp.Header.Set(xFromCache, "1")
+		return cachedResp, nil
+	}
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func refreshCache(bc *bugzillaCache, m prometheus.Gauge) {
+	concurrencyLimiter := make(chan string, 10)
+	logrus.WithField("cache_entries", len(bc.cache)).Info("Refreshing cache")
+	m.Set(float64(len(bc.cache)))
+	for url := range bc.cache {
+		concurrencyLimiter <- url
+		go func() {
+			url := <-concurrencyLimiter
```

Comment:
Why this over a semaphore? This is implicit, a semaphore explicit in intent

Reply:
Yeah - a semaphore is probably appropriate here. I'll change that